### PR TITLE
FIX: various revision history modal quirks

### PR DESCRIPTION
**Pull request by ZogStriP**
- FIX: properly scope category changes to what the current user can see
- UX: previous category is now highlighted in "red", new category is highlighted in "green"
- PERF: no need to serialize the categories
- FIX: properly track wiki
- FIX: properly track post_type (aka. Staff Color)
- FIX: properly track making a topic a PM
- FIX: never show the category changes when a topic is made a PM
- PERF: post_revision serializer is now more leaner (never includes title changes when post_number > 1, never includes user changes if there aren't any)
- UX: always sort the tags by name

Context - https://meta.discourse.org/t/-/304696

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/modal/history/revision.hbs [hunk 1]
```diff
@@ -35,17 +35,29 @@
...
+          {{d-icon "shield-alt" class=(if (eq @model.post_type_changes.current @site.post_types.moderator_action) "diff-del" "diff-ins")}}
         {{/if}}
 
-        {{#if @model.category_id_changes}}
-          {{html-safe @previousCategory}}
+        {{#if @model.archetype_changes}}
+          {{d-icon (if (eq @model.archetype_changes.current "private_message") "envelope" "comment")}}
+        {{/if}}
+
+        {{#if (and @model.category_id_changes (not @model.archetype_changes))}}
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/modal/history/revision.hbs:49]: We only show category changes when we haven't converted the topic into a PM.

File: app/assets/javascripts/discourse/app/components/modal/history/revision.hbs [hunk 2]
```diff
@@ -35,17 +35,29 @@
...
         {{/if}}
 
         {{#if @model.post_type_changes}}
-          <DisabledIcon @icon="shield-alt" @disabled={{@postTypeDisabled}} />
+          {{d-icon "shield-alt" class=(if (eq @model.post_type_changes.current @site.post_types.moderator_action) "diff-del" "diff-ins")}}
         {{/if}}
 
-        {{#if @model.category_id_changes}}
-          {{html-safe @previousCategory}}
+        {{#if @model.archetype_changes}}
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/modal/history/revision.hbs:45]: Show when we converted the topic to a PM or a public topic.

File: app/serializers/post_revision_serializer.rb [hunk 3]
```diff
@@ -25,11 +25,9 @@ class PostRevisionSerializer < ApplicationSerializer
              :title_changes,
              :user_changes,
              :tags_changes,
-             :wiki,
+             :category_id_changes,
              :can_edit
 
-  has_many :categories, serializer: CategoryBadgeSerializer, embed: :objects
```
Comment (ZogStriP) [inline app/serializers/post_revision_serializer.rb:31]: Those were not required. Just the category ids is enough.

File: app/serializers/post_revision_serializer.rb [hunk 4]
```diff
@@ -42,6 +40,7 @@ def self.add_compared_field(field)
   end
 
   add_compared_field :wiki
+  add_compared_field :post_type
```
Comment (ZogStriP) [inline app/serializers/post_revision_serializer.rb:43]: We needed that to track whenever a post gets a "staff color"

File: app/serializers/post_revision_serializer.rb [hunk 5]
```diff
@@ -25,11 +25,9 @@ class PostRevisionSerializer < ApplicationSerializer
              :title_changes,
              :user_changes,
              :tags_changes,
-             :wiki,
+             :category_id_changes,
```
Comment (ZogStriP) [inline app/serializers/post_revision_serializer.rb:28]: We don't need to know whether a post is currently a `wiki`. We're only interested in the changes. Which is already handled by the `add_compared_field :wiki` below.

But we do need to filter any category changes the current user can't see 🙈

File: app/serializers/post_revision_serializer.rb [hunk 6]
```diff
@@ -139,10 +135,13 @@ def title_changes
     { inline: diff.inline_html, side_by_side: diff.side_by_side_html }
   end
 
+  def include_title_changes?
```
Comment (ZogStriP) [inline app/serializers/post_revision_serializer.rb:138]: Let's not serialize the title changes on "replies"

**Review by pmusaraj — APPROVED** [review]
Looks good!

**Pull request merged**